force_bdss/core_mco_driver.py

```diff
@@ -30,5 +47,50 @@ def application_started(self):
 
         mco_model = workflow.mco
         mco_factory = mco_model.factory
-        mco = mco_factory.create_optimizer()
-        mco.run(mco_model)
+        return mco_factory.create_optimizer()
+
+    @on_trait_change("mco:event")
+    def _handle_mco_event(self, event):
+        for listener in self.listeners:
+            try:
+                listener.deliver(event)
+            except Exception as e:
+                log.error(
+                    "Exception while delivering to listener {}: {}".format(
+                        listener.__class__.__name__,
+                        str(e)
+                    ))
+                self._finalize_listener(listener)
+                self.listeners.remove(listener)
+
+    def _listeners_default(self):
+        listeners = []
+
+        for nl_model in self.workflow.notification_listeners:
+            factory = nl_model.factory
+            try:
+                listener = factory.create_listener()
+                listener.initialize(nl_model)
+            except Exception as e:
+                log.error(
+                    "Failed to create or initialize "
+                    "listener with id {}: {}".format(
+                        factory.id, str(e)))
+            else:
+                listeners.append(listener)
+
+        return listeners
+
+    def _finalize_listener(self, listener):
+        """Helper method. Finalizes a listener and handles possible
+        exceptions. it does _not_ remove the listener from the listener
+        list.
+        """
+        try:
+            listener.finalize()
+        except Exception as e:
+            log.error(
+                "Exception while finalizing listener {}: {}".format(
+                    listener.__class__.__name__,
+                    str(e)
+                ))
```

Comment:
Should we not remove the listener from the list of listeners if it failed to finalize ? At this point we can not do anything for this listener..